cnm.go

```diff
@@ -16,29 +16,140 @@
 
 package virtcontainers
 
+import (
+	"net"
+
+	"github.com/01org/ciao/ssntp/uuid"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/types"
+)
+
 // cnm is a network implementation for the CNM plugin.
 type cnm struct {
 	config NetworkConfig
 }
 
+func (n *cnm) createResult(addrs []net.Addr) (types.Result, error) {
+	var res types.Result
+
+	return res, nil
+}
+
+func (n *cnm) createEndpointsFromScan() ([]Endpoint, error) {
+	var endpoints []Endpoint
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return []Endpoint{}, err
+	}
+
+	uniqueID := uuid.Generate().String()
+
+	for i, iface := range ifaces {
+		var endpoint Endpoint
+
+		addrs, err := iface.Addrs()
+		if err != nil {
+			return []Endpoint{}, err
+		}
+
+		if iface.Name == "lo" {
+			continue
+		} else {
+			endpoint = createNetworkEndpoint(i, uniqueID, iface.Name)
+		}
+
+		endpoint.Properties, err = n.createResult(addrs)
+		if err != nil {
+			return []Endpoint{}, err
+		}
+
+		endpoints = append(endpoints, endpoint)
+	}
+
+	return endpoints, nil
+}
+
 // init initializes the network, setting a new network namespace for the CNM network.
 func (n *cnm) init(config *NetworkConfig) error {
+	if config.NetNSPath == "" {
+		path, err := createNetNS()
+		if err != nil {
+			return err
+		}
+
+		config.NetNSPath = path
+	}
+
 	return nil
 }
 
 // join switches the current process to the specified network namespace
 // for the CNM network.
 func (n *cnm) join(networkNSPath string) error {
+	err := setNetNS(networkNSPath)
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
 
 // add adds all needed interfaces inside the network namespace for the CNM network.
 func (n *cnm) add(pod Pod, config NetworkConfig) (NetworkNamespace, error) {
-	return NetworkNamespace{}, nil
+	endpoints, err := n.createEndpointsFromScan()
+	if err != nil {
+		return NetworkNamespace{}, err
+	}
+
+	networkNS := NetworkNamespace{
+		NetNsPath: config.NetNSPath,
+		Endpoints: endpoints,
+	}
+
+	err = doNetNS(networkNS.NetNsPath, func(_ ns.NetNS) error {
+		for _, endpoint := range networkNS.Endpoints {
+			err := bridgeNetworkPair(endpoint.NetPair)
```

Comment:
@sboeuf @mcastelino Yes, let's postpone this generalization to the SRIOV support.